### ISSUE TITLE
exception while downloading while trying "git clone"

### ISSUE THREAD

**Issue by Armagedoom** [anishathalye/git-remote-dropbox#67]
When trying to clone the repo from dropbox `git clone dropbox:///raspi_home/servers_foldere` I get:

```
error: exception while downloading: ApiError('2a66a732c4964fa2316abde6590c7c07', DownloadError(u'path', LookupError(u'not_found', None)))
```

I think it shouldnt happen because:
* I can push to that repository from within the original folder
* If I try to pull another (non existing) folder, i get the error:
```
info: repository is empty
warning: You appear to have cloned an empty repository.
```

So the problem is Im not being able to clone the existing repository which otherwise is working properly.

I tried debugging and when I launch the script, at a point it stops to parse a command from stdin, and when I input "clone" i get the error:

`error: unsupported operation: clone`

Is cloning supported? How should I clone a repo I created using `git-remote-dropbox`?

I have time this week so I could contribute if someone gives me clues on what to do.
Thank you!

**Comment by anishathalye** [comment]
See #58?

tl;dr it appears as if the file is missing from your Dropbox. It would be great if we could figure out exactly how/why that happened.

**Comment by Armagedoom** [comment]
Thank you very much.
First of all, i thought this would happen with all my repos using `git-remote-dropbox`, but turns out it only fails in one, and the rest work perfectly.

Here is all the info I have gathered about the topic so far:

* I did some pushes to the repo and now when i try to clone it, it starts to "receive objects" and then fails with an error on a random file each time:

```
Receiving objects:  15% (66/431)error: exception while downloading: ApiError('3bxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0', DownloadError(u'path', LookupError(u'not_found', None)))
```

In that example the error was in `15% (66/431)` but if i try again it will give a random number every time:
`79% (354/447)`
`46% (16/35)`
`9% (28/317)`

My feeling is this had something to do with the way the repository was first created and pushed, regarding to the branches (no master branch manually created, but anyway, then worked with that...)

Then, after trying to clone several times and stopping at random values, I got this error which seems unrelated to `git-remote-dropbox`:

```
Cloning into 'servers_folder'...
Receiving objects:  79% (354/447)error: exception while downloading: ApiError('adfc40a53bbe73ac0b56dfb8ecb90601', DownloadError(u'path', LookupError(u'not_found', None)))
Exception in thread Thread-16 (most likely raised during interpreter shutdown):
 Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_innerException in thread Thread-15 (most likely raised during interpreter shutdown):
  File "/usr/lib/python2.7/threading.py", line 763, in run
Exception in thread Thread-10 (most likely raised during interpreter shutdown):  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 295, in __call__
  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 540, in _downloadException in thread Thread-6 (most likely raised during interpreter shutdown):Exception in thread Thread-9 (most likely raised during interpreter shutdown):


  File "/usr/lib/python2.7/multiprocessing/queues.py", line 117, in get
<type 'exceptions.EOFError'>:

Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 763, in run
  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 295, in __call__
  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 540, in _download
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 117, in get
```

After a few more trials, I got another error:

```
 Exception in thread Thread-16 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 763, in run
  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 295, in __call__
  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 543, in _download
  File "/usr/local/lib/python2.7/dist-packages/git_remote_dropbox/__init__.py", line 510, in _get_file
  File "/usr/local/lib/python2.7/dist-packages/dropbox/base.py", line 416, in files_download
  File "/usr/local/lib/python2.7/dist-packages/dropbox/dropbox.py", line 234, in request
  File "/usr/local/lib/python2.7/dist-packages/dropbox/dropbox.py", line 325, in request_json_string_with_retry
  File "/usr/local/lib/python2.7/dist-packages/dropbox/dropbox.py", line 409, in request_json_string
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 549, in post
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 502, in request
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 612, in send
  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 489, in send
<type 'exceptions.AttributeError'>: 'NoneType' object has no attribute 'error'
```

Tomorrow I will try to include within the code a "print" of the file trying to be cloned, so I can see which specific file is causing the trouble and see the differences between that file in the local and in the dropbox.

In case I reach nowhere, is there any data or info you would like to keep in order to try and track this issue before I turn to the solution mentioned in #58?

Thanks again for your feedback.

**Comment by anishathalye** [comment]
One thing to check -- are the relevant files actually missing from your Dropbox? (see [here](https://github.com/anishathalye/git-remote-dropbox/issues/58#issuecomment-294566438) for details on the naming convention)

**Comment by Armagedoom** [comment]
The last file on the debug list exists, and the refference in the error message "892bb9d060fb781f37378aa648438a21" does not exist in either of the repos.

![image](https://user-images.githubusercontent.com/4046326/29552848-3d5578d4-8719-11e7-93aa-297c58136dc8.png)

Im pushing the repo into a new repo within my dropbox to see if the error stays or not.
If not, this should be related to how the repo was initialized and the order in which the original branches / commits / push were made.

Will come back with any findings.

**Comment by Armagedoom** [comment]
The newly created repository (just by pushing the same non-working one) works perfectly.
The remote repo has much more files within /objects/ than the failing remote repo.
 Git clone finished without trouble.
The files in the local repo are exactly the same, since I used the same repo just changing the "url" in the `.git/config` file.
Changed the URL again, did git push, and tried git clone again, and it fails, even though it worked perfectly with the freshly pushed repo.
Im a bit clueless into what to check next. 
The problem is "solved" for me,  but I will keep the failing repo in case you want me to test anything.
Thanks!

**Comment by Armagedoom** [comment]
Wiped Dropbox repository, pushed the repo, got the error again when cloning it.
Going to try that again see if it is broken again.

**Comment by Armagedoom** [comment]
Removed the .git folder, and re-built the git with:

``` 
git init
git remote add origin "dropbox:///python/book-testing-with-pytest"
git checkout -b master
git add *
git commit -m "first commit"
git push origin master
```

And now its clonable without error.
I believe the problem lays within the order of master branch creation / first commit / push.

If the problem comes back this means I am wrong, and I would report in such case.

If you need some tests done, I can try to recover the wrong repo files, after all, they are in dropbox

**Comment by anishathalye** [comment]
Thanks for taking the time to try to look into this.

Closing because of inactivity. If you have any new information, please reopen the issue.

**Comment by tianboxi** [comment]
I'm encountering the same issue while trying to clone a repository using:
`git clone "dropbox:///code.git"`
I get error:
`Cloning into 'code'...
Receiving objects:  13% (2199/16827)error: exception while downloading: HTTPSConnectionPool(host='content.dropboxapi.com', port=443): Max retries exceeded with url: /2/files/download (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x110796790>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))`

My mac version is macos Mojave and the repo was originally created by push from Ubuntu using git-remote-dropbox

Thanks for any help!

**Comment by anishathalye** [comment]
Looks like a network issue (perhaps a transient error?). If you try again, what happens?

**Comment by ChausseBenjamin** [comment]
@tianboxi I've had this error popup also... (My repo weighed 6gb) What I did to help out was run it in debug. In your case, that would be:
`git clone "dropbox:///code.git" -v`

**Comment by tianboxi** [comment]
@ChausseBenjamin Thanks for the advice. 

It seems that the problem is related to the size of the repository. I tried with my smaller repo (~100Mb) and it cloned with no problem. When I clone my larger repo (~2.7gb), the same error appears every time at a random percentage when fetching objects as @Armagedoom described originally.

I kind of solved the problem by download the .git folder from Dropbox, manually set up the repo and add the remote. Then everything function normally and I can push or pull with git-remote-dropbox.

**Comment by ChausseBenjamin** [comment]
Great to hear!

**Comment by voronamanga** [comment]
Hi,
I had the same issue with a 1.0 GiB repo, not when launching git clone but git pull into an empty repo. I know this thread is pretty old, but it fits my problem and I found no other relevant one. I would just like to point out that, thanks to @tianboxi I got more confident in the kind of manipulation he/she was mentioning ; so I completed .git directory by downloading directly from Dropbox. After that, running git pull completed successfully and git status showed up a clean working tree. I finally copied/pasted the actual files (-> working directory) and committed. It seems to be working. 
I would like to say in addition that git-remote-dropbox appears to me like an important piece of software for small projects with few developpers, since using Dropbox as server is really interesting in thoses cases - as of course large teams probably use dedicated servers. So I think it can have a nice future.

**Comment by mohorko** [comment]
Hi,

I have been having similar issues on some of my repos for some time now. Until now, I simply manually deleted the dropbox repo and re-created it from local source. But this is getting annoying - so today I went googling for a solution and it brough me here ... just to report that this is still an open issue.
My repos are small (less then 100MB, some only a couple of MB), so the repo size is not a factor for this error.


My dropbox git repos are also my backup incase I loose/break my laptop - so I am worried about this "_remote repo corruption_" because ... what if I do not have a local source to rebuild the repo from?
